chat-android/src/main/java/com/ably/chat/Typing.kt

```diff
@@ -163,82 +187,132 @@ internal class DefaultTyping(
     override suspend fun get(): Set<String> {
         logger.trace("DefaultTyping.get()")
         room.ensureAttached(logger) // CHA-T2d, CHA-T2c, CHA-T2g
-        return channelWrapper.presence.getCoroutine().map { it.clientId }.toSet()
+        return currentlyTypingMembers
     }
 
     override suspend fun start() {
         logger.trace("DefaultTyping.start()")
-
-        typingScope.launch {
-            // If the user is already typing, reset the timer
-            if (typingJob != null) {
-                logger.debug("DefaultTyping.start(); already typing, resetting timer")
-                typingJob?.cancel()
-                startTypingTimer()
-            } else {
-                startTypingTimer()
-                room.ensureAttached(logger) // CHA-T4a1, CHA-T4a3, CHA-T4a4
-                channelWrapper.presence.enterClientCoroutine(room.clientId)
+        typingScope.async {
+            room.ensureAttached(logger) // CHA-T4a1, CHA-T4a3, CHA-T4a4
+            // If the user is already typing, return
+            if (typingInProgressJob?.isActive == true) {
+                logger.warn("DefaultTyping.start(); already typing, so it's a no-op")
+                return@async
```

Comment:
probably should be `return` here, otherwise there is no logic to prevent sending start typing event

Reply:
I didn't get

Reply:
you unconditionally call `startHeartbeatTimer()` and `sendTyping()` every time

Reply:
Btw, just checked, new spec item [CHA-T14] has been added as a part of latest commit https://github.com/ably/specification/pull/277/commits/e8d46fdd1442c65ae0aea16f4f4e575aa0d74906. 
This says to wait for previous call to complete before new one starts. 
Also, as per CHA-T4c, If typing is already in progress, The client must not send a @typing.started@ event

Reply:
Updated